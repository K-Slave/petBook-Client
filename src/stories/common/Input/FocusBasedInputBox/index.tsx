import React, { PropsWithChildren, useMemo } from "react";
import inputEventHelperMethod from "@lib/modules/login/inputEventHelperMethod";
import { CommonInputProps } from "../CommonInput/CommonInput";
import {
  FocusBasedAlginCenterLabel,
  FocusBasedBgSpan,
  FocusBasedInputBoxElem,
} from "./style";

// 포커스 기반 Input Box
// Label 태그로 감싸져 있음
export interface FocusBasedInputBoxProps extends CommonInputProps {
  inputRef?: React.RefObject<HTMLInputElement>;
  onChange?: (e: React.ChangeEvent<HTMLInputElement>) => void;
  isTyping?: boolean;
  bgUrl?: string;
  bgWidth?: string;
  bgHeight?: string;
}

const FocusBasedInputBox = (
  props: PropsWithChildren<FocusBasedInputBoxProps>
) => {
  const { children, type, bgUrl, bgWidth, bgHeight, register, onChange } =
    props;
  const inputProps = { ...props };
  delete inputProps.children;

  const onBlur: React.FocusEventHandler<HTMLInputElement> = (e) => {
    const handler = new inputEventHelperMethod(e);

    handler.setFocused("remove");
    if (e.target.validity.valid) {
      handler.setValid("add");
    } else if (!e.target.validity.valid && e.target.value.length > 0) {
      e.target.classList.add("invalid");
      handler.setInvalid("add");
    }
  };

  const onFocus: React.FocusEventHandler<HTMLInputElement> = (e) => {
    const handler = new inputEventHelperMethod(e);

    handler.setFocused("add");
    handler.setValid("remove");
    handler.setInvalid("remove");
  };

  const dataType = useMemo(() => type, []);

  return (
    <FocusBasedAlginCenterLabel data-type={dataType}>
      {children}
      {bgUrl && (
        <FocusBasedBgSpan
          url={bgUrl || ""}
          width={bgWidth || "1rem"}
          height={bgHeight || "1rem"}
        />
      )}
      <FocusBasedInputBoxElem
        {...inputProps}
        type={type}
        onChange={register ? register.onChange : onChange}
        onBlur={register ? register.onBlur : onBlur}
        onFocus={onFocus}
        bgWidth={bgWidth}
        register={register}
        placeholderColor="var(--black_06)"
      />
    </FocusBasedAlginCenterLabel>
  );
};

export default FocusBasedInputBox;
