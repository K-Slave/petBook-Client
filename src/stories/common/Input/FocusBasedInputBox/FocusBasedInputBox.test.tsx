import { render, fireEvent } from "@testing-library/react";
import FocusBasedInputBox from ".";
import React, { useState } from "react";
import renderer from "react-test-renderer";

const TestComponent = () => {
  const [value, setValue] = useState("test");
  const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setValue(e.target.value);
  };
  return (
    <FocusBasedInputBox
      type="text"
      value={value}
      onChange={onChange}
      bgUrl=""
      bgHeight=""
      bgWidth=""
    />
  );
};

describe("FocusBasedInputBox", () => {
  it("should render without errors", () => {
    const { container } = render(<TestComponent />);

    expect(container).toBeInTheDocument();
  });

  test("matches snapshot", () => {
    const tree = renderer.create(<TestComponent />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('adds "valid" class on blur if input is valid', () => {
    render(<TestComponent />);
    const input = document.querySelector("input");

    if (input) {
      // Fire blur event
      fireEvent.blur(input);
      expect(input).toHaveClass("valid");
    }
  });

  test('adds "invalid" class on blur if input is invalid', () => {
    render(<TestComponent />);
    const input = document.querySelector("input");

    // Make input invalid
    input!.setCustomValidity("Invalid");

    // Fire blur event
    fireEvent.blur(input!);
    expect(input).toHaveClass("invalid");
  });

  test('removes "valid" and "invalid" classes on focus', () => {
    render(<TestComponent />);
    const input = document.querySelector("input");

    // Add valid and invalid classes
    input!.classList.add("valid");
    input!.classList.add("invalid");

    // Fire focus event
    fireEvent.focus(input!);
    expect(input).not.toHaveClass("valid");
    expect(input).not.toHaveClass("invalid");
  });
});
