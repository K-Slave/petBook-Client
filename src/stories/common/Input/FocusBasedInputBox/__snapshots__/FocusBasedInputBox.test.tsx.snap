// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FocusBasedInputBox matches snapshot 1`] = `
<label
  className="sc-eCYdqJ kkuEQo sc-gKXOVf foouDq"
>
  <input
    className="sc-jSMfEi btQVOu sc-ftvSup cYobvy"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="내용을 입력해주세요."
    type="text"
    value="test"
  />
</label>
`;
